Fastify-2 blog post (#98)
val-town/val-town-blog: src/content/blog/fastify-2.mdx

---
title: "Moving to fastify pt. 2: Docs, an SDK and AI tools"
description: |
  Taking advantage of our typed REST API to build a platform around
  Val Town.

> pubDate: July 12, 2024

Reviewer: Reminder to update :)

author: Tom MacWright
---

> The last time, I [wrote about how we were switching from express to fastify][last_post]
> for Val Town’s REST API. To recap, Fastify gives us a way to strictly

Reviewer: ```suggestion
We recently [switched from express to fastify][last_post]
for Val Town’s REST API. Fastify gives us a way to strictly
```

request and response bodies aligning to a specific schema. This is a big
win for robustness, gives us great type-safety with TypeScript, and lets us
generate an [OpenAPI Specification][openapi] directly from our server code.

But typesafety and robustness don't bring home the bacon. What do you,
the user, get from our time spent refactoring?

### Nice new REST API documentation

![Val Town REST API Documentation](./fastify-2/docs.png)

First thing, we get **way better REST API documentation**. It has example
requests, with everything from cURL to Python clients. It describes the precise
shapes of the objects you'll get back. You can even test APIs right in
the API docs. Hats off to [Scalar](https://scalar.com), whose
[open source package](https://github.com/scalar/scalar) is powering this.

The docs are based on our OpenAPI spec, which is emitted from Fastify and
pulled from our server's routes. From there on, setting up nice documentation
is basically as simple as this:

```html
<script id="api-reference" data-url="./openapi.documented.json"></script>
<script src="https://cdn.jsdelivr.net/npm/@scalar/api-reference"></script>
```

### A TypeScript SDK

> Are you not entertained? How about [a new TypeScript SDK](https://github.com/val-town/sdk)?

Reviewer: ```suggestion
How about [a new TypeScript SDK](https://github.com/val-town/sdk)?
```

Reviewer: (I didn't find the joke funny, I guess?)

```ts title="example.ts"
// Import the SDK
import ValTown from "@valtown/sdk";
// Initialize it
const valTown = new ValTown();
// Use it
console.log(await valTown.me.profile.retrieve());
```

The SDK covers all of our major API methods, and has robust TypeScript
types for everything. Methods in our API that are paginated, like listing
comments, vals, or search results, are implemented as async iterators
in the SDK so you don't even have to implement any pagination logic.

> Just install it.

Reviewer: I think we should prioritize showing off its use in Val Town first and foremost (link to a val or two?) and then say of course it works in all the other js runtimes

```bash
npm install @valtown/sdk
```

The SDK works in Deno, of course, but also in Node.js, Bun, and other major
runtimes. We're generating it with tools from our friends at [Stainless](https://www.stainlessapi.com/), and their system keeps it updated
so that every time we update the REST API, we can quickly roll a new release
of the SDK.

### Our AI, Townie, can now call our REST API

![Townie screenshot](./fastify-2/townie.png)

[Townie](https://www.val.town/townie) is Val Town's bot - it started as an
experiment to see if we could coax an LLM into writing code that fit
Val Town's runtime better. It has a long way to go before it reaches
sentience, fortunately/unfortunately, but it can already write some pretty
decent code.

But it couldn't really interact with Val Town until now. Townie just got
the ability to call our API directly, which means that it can do things like
creating vals, searching for existing code, updating readmes – anything
that the API can do, Townie can do too.
To be safe, Townie will run `GET` requests against the API freely, but asks
for permission before running `POST`, `PUT`, or `DELETE` requests.

> 

Reviewer: A video here could be awesome, but not necessary

Townie knows the whole layout of the API because OpenAPI 3.1 is
[compatible with JSON Schema][oa31], and
the [OpenAI Function calling][calling]
system is based on JSON Schema-defined parameters. So we can fairly smoothly
translate from our OpenAPI specification for our REST API, into a definition
of functions that OpenAI can call – along with LLM-friendly descriptions
for every function, parameter, and return value – and voila, the robot
can use the machine. Avoiding OpenAPI-OpenAI mixup typos was one of the more
mentally challenging parts of implementing this.

> 

Reviewer: Feels a bit weird to end on that note. Some ideas:

* any final thoughts on Fastify/OpenAPI/OpenAI, migrations, etc
* Tips for others
* Future directions

![Townie screenshot](./fastify-2/townie.png)

[Townie](https://www.val.town/townie) is Val Town's bot - it started as an
experiment to see if we could coax an LLM into writing code that fit
Val Town's runtime better. It has a long way to go before it reaches
sentience, fortunately/unfortunately, but it can already write some pretty
decent code.

But it couldn't really interact with Val Town until now. Townie just got
the ability to call our API directly, which means that it can do things like
creating vals, searching for existing code, updating readmes – anything
that the API can do, Townie can do too.
To be safe, Townie will run `GET` requests against the API freely, but asks
for permission before running `POST`, `PUT`, or `DELETE` requests.

Townie knows the whole layout of the API because OpenAPI 3.1 is
[compatible with JSON Schema][oa31], and
the [OpenAI Function calling][calling]
system is based on JSON Schema-defined parameters. So we can fairly smoothly
translate from our OpenAPI specification for our REST API, into a definition
of functions that OpenAI can call – along with LLM-friendly descriptions
for every function, parameter, and return value – and voila, the robot

> can use the machine. Avoiding OpenAPI-OpenAI mixup typos was one of the more

Reviewer 2: > and voila, the robot can use the machine.

I really like this line

Author: Thx!
